### PR TITLE
Address bugs affecting UTF-8 input files

### LINKED ISSUE #7042: EPjson input file format does not support diacritics
Issue overview
--------------
Energyplus fails silently when running an epJSON file with diacritics ([example file failing with 1 'é', working if replaced by 'e'](https://github.com/NREL/EnergyPlus/files/2534475/N4.epJSON.txt)). It might not be a bug but adding a meaningful error message would be nice, especially considering that diacritics works well in IDF files.

### Details
Some additional details for this issue (if relevant):
 - Platform : windows 10
 - Version of EnergyPlus : 8.9.0-eba93e8e1b

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #7043: EnergyPlus Not Supporting UTF-8
Issue overview
--------------
EnergyPlus no longer seems to support UTF-8 encoding as it did before breaking it's use for a lot of international applications. I tested using various UTF-8 characters, such as 模型, and found that from EnergyPlus 8.9 onwards the files don't run. EnergyPlus 8.8 worked, but 8.9 and 9.0.1 did not work.

Some additional details for this issue (if relevant):
 - Windows 10 + Mac OS Sierra
 - 8.9 and 9.0.1

### PR DESCRIPTION
**Pull request by mbadams5**
Pull request overview
---------------------
Fixes #7043 
Fixes #7042

This pull request addresses a few bugs related to UTF-8 input files. EnergyPlus now has equivalent behavior as far as I can tell compared to EP8.8 and lower. In fact, it now has stricter requirements than previous EnergyPlus versions since it requires UTF-8 encoding. Previously, EnergyPlus was blissfully unaware of input file encoding and only cared about extended ASCII for upper/lower casing strings.

If an IDF has a different encoding than UTF-8, the current default behavior will replace non-conforming UTF8 bytes with U+FFFD. The three JSON library options (strict, replace, ignore) for dumping a JSON file to disk are as follows:
* [json.exception.type_error.316] invalid UTF-8 byte at index 2: 0xA9 string with replaced invalid characters: "ä�ü" string with ignored invalid characters: "äü"
* string with replaced invalid characters: "ä�ü"
* string with ignored invalid characters: "äü"

This pull request also fixes a bug that now allows the HTML report to show unicode strings instead of the literal characters.
Additionally, this PR provides error messaging where there was none previously (usually during parsing errors).

I created unit tests and added an integration test that have UTF-8 characters in them.

The UTF-8 characters are visible in ESO, EIO, HTML, and SQL files as shown in screenshots below.
ESO
EIO
HTML
SQL

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Required documentation updates?
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

File: src/EnergyPlus/GroundHeatExchangers.cc [hunk 1]
```diff
@@ -597,9 +598,9 @@ namespace GroundHeatExchangers {
 
         // save data for later
         if (!DataSystemVariables::DisableGLHECaching) {
-            myCacheData["Response Factors"]["time"] = myRespFactors->time;
-            myCacheData["Response Factors"]["LNTTS"] = myRespFactors->LNTTS;
-            myCacheData["Response Factors"]["GFNC"] = myRespFactors->GFNC;
+            myCacheData["Response Factors"]["time"] = std::vector<Real64>(myRespFactors->time.begin(), myRespFactors->time.end());
+            myCacheData["Response Factors"]["LNTTS"] = std::vector<Real64>(myRespFactors->LNTTS.begin(), myRespFactors->LNTTS.end());
+            myCacheData["Response Factors"]["GFNC"] = std::vector<Real64>(myRespFactors->GFNC.begin(), myRespFactors->GFNC.end());
```
Comment (Myoldmopar) [inline src/EnergyPlus/GroundHeatExchangers.cc:603]: Random fix here, but good catch.  FYI @mitchute.

File: src/EnergyPlus/InputProcessing/IdfParser.cc [hunk 2]
```diff
@@ -659,14 +659,11 @@ IdfParser::Token IdfParser::next_token(std::string const &idf, size_t &index)
...
     default:
-        static std::string const search_chars("-:.#/\\[]{}_@$%^&*()|+=<>?'\"~");
         static std::string const numeric(".-+0123456789");
         if (numeric.find_first_of(c) != std::string::npos) {
             return Token::NUMBER;
-        } else if (isalnum(c) || (std::string::npos != search_chars.find_first_of(c))) {
-            return Token::STRING;
         }
-        break;
+        return Token::STRING;
```
Comment (Myoldmopar) [inline src/EnergyPlus/InputProcessing/IdfParser.cc:666]: Nice.  Simpler.

File: src/EnergyPlus/OutputReportTabular.cc [hunk 3]
```diff
@@ -3577,6 +3577,7 @@ namespace OutputReportTabular {
                                << '\n';
                     tbl_stream << " - EnergyPlus</title>\n";
                     tbl_stream << "</head>\n";
+                    tbl_stream << "<meta http-equiv=\"Content-Type\" content=\"text/html; charset=utf-8\">\n";
```
Comment (Myoldmopar) [inline src/EnergyPlus/OutputReportTabular.cc:3580]: :+1:

File: testfiles/CMakeLists.txt [hunk 4]
```diff
@@ -16,6 +16,7 @@ ADD_SIMULATION_TEST(IDF_FILE 1ZoneUncontrolledAnnualOutputs.idf EPW_FILE USA_CO_
 ADD_SIMULATION_TEST(IDF_FILE 1ZoneUncontrolledCondFDWithVariableKat24C.idf EPW_FILE USA_CO_Golden-NREL.724666_TMY3.epw)
 ADD_SIMULATION_TEST(IDF_FILE 1ZoneUncontrolledFourAlgorithms.idf EPW_FILE 94810-1956.epw)
 ADD_SIMULATION_TEST(IDF_FILE 1ZoneUncontrolledResLayers.idf EPW_FILE 94810-1956.epw)
+ADD_SIMULATION_TEST(IDF_FILE 1ZoneUncontrolledUTF8.idf EPW_FILE USA_CO_Golden-NREL.724666_TMY3.epw)
```
Comment (Myoldmopar) [inline testfiles/CMakeLists.txt:19]: :+1 :+1:

File: tst/EnergyPlus/unit/InputProcessor.unit.cc [hunk 5]
```diff
@@ -374,6 +374,186 @@ TEST_F(InputProcessorFixture, parse_empty_fields)
...
+      "}}"
+    );
+
+    ASSERT_TRUE(process_idf(idf));
+    json &epJSON = getEpJSON();
+
+    auto const input_file = epJSON.dump(-1, ' ', false, json::error_handler_t::replace);
+
+    EXPECT_EQ( expected, input_file );
+}
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/InputProcessor.unit.cc:555]: Awesome.

**Review by Myoldmopar — APPROVED** [review]
Super thorough.  It's awesome to see the small fixes here and there that enable this to get back to the way it was prior to 8.9.  Great new unit tests and new example file.  I'm glad to see the SQLite changes were super minimal.

File: src/EnergyPlus/GroundHeatExchangers.cc (hunk 1 above)
Comment (mbadams5) [reply src/EnergyPlus/GroundHeatExchangers.cc:603]: Not entirely random. This gave a build error with a newer version of the JSON library.

**Comment by Myoldmopar** [comment]
Is [this change](http://energyplus.s3-website-us-east-1.amazonaws.com/regressions/2018-11/d4e72c89b91b70bf6662bed2764514ad6410e3eb-49113e23dd58403413812cf28f1dddfb934bbde5/PlantApplicationsGuide_Example2/x86_64-MacOS-10.9-clang/eplusout.err.diff.html) in the error file output on Mac is expected?

**Comment by mbadams5** [comment]
No, but I don't know if it is related to these changes or not. I'm seeing non-determinism when running that file regarding that line in the error file. I got 6 Veterans Day, 2 Columbus Day, 1 Presidents Day, and 0 Memorial Day.

**Comment by mbadams5** [comment]
@Myoldmopar I see why there is the non-determinism. I am using a [json pointer as the key in a map](https://github.com/NREL/EnergyPlus/blob/92dc6f0b3ccea8d1dbe6fd6f8013c39edcc7aad8/src/EnergyPlus/InputProcessing/InputProcessor.hh#L238), where it is getting added [here](https://github.com/NREL/EnergyPlus/blob/92dc6f0b3ccea8d1dbe6fd6f8013c39edcc7aad8/src/EnergyPlus/InputProcessing/InputProcessor.cc#L208-L209).

I am surprised none of the other platforms have this non-determinism and that it hasn't been seen previously either. There must be a change in the JSON library regarding the JSON pointers.

If we want to address this then I'll need to think about how to fix this since I used a pointer on purpose (for space and speed considerations).

**Comment by Myoldmopar** [comment]
So we have seen sporadic non-determinism on things like this before, but I don't recall if it was every platform or just Mac.  Anyway, I think we _do_ want to address it, so go ahead and take some time to think about it and let me know if you want me to do anything.

**Comment by Myoldmopar** [comment]
EIO diffs are unrelated due to changes in develop.  This is great!  Thanks for the rapid fix, @mbadams5, I'm merging now.

**Pull request merged**